RaftMongoWithRaftReconfig.tla

```diff
@@ -0,0 +1,276 @@
+VARIABLE commitPoint
+
+\* The server's state (Follower, Candidate, or Leader).
+VARIABLE state
+
+electionVars == <<currentTerm, state>>
+serverVars == <<electionVars, commitPoint>>
+
+\* A Sequence of log entries. The index into this sequence is the index of the
+\* log entry. Unfortunately, the Sequence module defines Head(s) as the entry
+\* with index 1, so be careful not to use that!
+VARIABLE log
+logVars == <<log>>
+
+\* End of per server variables.
+----
+
+\* All variables; used for stuttering (asserting state hasn't changed).
+vars == <<serverVars, logVars, configs>>
+
+----
+\* Helpers
+
+\* The term of the last entry in a log, or 0 if the log is empty.
+GetTerm(xlog, index) == IF index = 0 THEN 0 ELSE xlog[index].term
+LogTerm(i, index) == GetTerm(log[i], index)
+LastTerm(xlog) == GetTerm(xlog, Len(xlog))
+
+\* Return the minimum value from a set, or undefined if the set is empty.
+Min(s) == CHOOSE x \in s : \A y \in s : x <= y
+\* Return the maximum value from a set, or undefined if the set is empty.
+Max(s) == CHOOSE x \in s : \A y \in s : x >= y
+
+\* The config version in the node's last entry.
+GetConfigVersion(i) == log[i][Len(log[i])].configVersion
+
+\* Gets the node's first entry with a given config version.
+GetConfigEntry(i, configVersion) == LET configEntries == {index \in 1..Len(log[i]) : 
+                                                            log[i][index].configVersion = configVersion}
+                                    IN Min(configEntries)
+
+\* The servers that are in the same config as i.
+ServerViewOn(i) == configs[GetConfigVersion(i)]
+
+\* The set of all quorums. This just calculates simple majorities, but the only
+\* important property is that every quorum overlaps with every other.
+Quorum(me) == {sub \in SUBSET(ServerViewOn(me)) : Cardinality(sub) * 2 > Cardinality(ServerViewOn(me))}
+
+----
+\* Define initial values for all variables
+InitServerVars == /\ currentTerm = [i \in Server |-> 0]
+                  /\ state       = [i \in Server |-> Follower]
+                  /\ commitPoint = [i \in Server |-> [term |-> 0, index |-> 0]]
+InitLogVars == /\ log         = [i \in Server |-> << [term |-> 0, configVersion |-> 1] >>]
+InitConfigs == configs = << Server >>
+Init == /\ InitServerVars
+        /\ InitLogVars
+        /\ InitConfigs
+
+----
+\* Message handlers
+\* i = recipient, j = sender, m = message
+
+AppendOplog(i, j) ==
+    /\ state[i] = Follower  \* Disable primary catchup and draining
+    /\ j \in ServerViewOn(i)  \* j is in the config of i.
+    /\ Len(log[i]) < Len(log[j])
+    /\ LastTerm(log[i]) = LogTerm(j, Len(log[i]))
+    /\ log' = [log EXCEPT ![i] = Append(log[i], log[j][Len(log[i]) + 1])]
+    /\ currentTerm' = [currentTerm EXCEPT ![i] = IF LogTerm(j, Len(log[i]) + 1) > currentTerm[i] 
+                                                 THEN LogTerm(j, Len(log[i]) + 1) 
+                                                 ELSE currentTerm[i]]
```

Comment:
My thought process for term propagation was to make as few restrictions as possible for how this is implemented in the code. I wanted to avoid masking a bug because, for example, the spec assumes the term is always propagated on rollback but the code doesn't guarantee it. This led me to the rules: (1) The term can propagate at any time (i.e. by heartbeat), and (2) The term need not propagate in every communication between nodes. This clarified for me when term propagation is actually required: (1) when appending an oplog entry with a higher term, (2) when voting in an election, and (3) when advancing the majority commit point. What do you think about this approach?

Reply:
Re "For (3), it doesn't appear we update terms at all when advancing the majority commit point.":
Sorry, I was unclear. We don't update terms when advancing the majority commit point, but it's necessary that terms are sent as part of acknowledging writes, so we can check none of the acknowledgers have a higher term.

Re "Whether or not we propagated that term to more nodes via AppendOplog actions seems irrelevant, though.":
Yes, you're correct, it's not necessary to propagate terms as part of AppendOplog. It just felt weird not to, since the node clearly knows of the higher term. I'll remove this term update.

Re "Maybe you could explain your reasoning behind why you changed that condition changed?"
I was trying to model how this property is enforced in the server, which is that if we hear of a higher term, then we step down, and we don't acknowledge the write as majority committed. I think that if we don't include the term update in AppendOplog, then the <= is necessary because the acknowledgers might not have the higher term.